include/smp/lock.h

```diff
@@ -22,86 +22,48 @@
 typedef enum {
     CLHState_Granted = 0,
     CLHState_Pending
-} clh_qnode_state_t;
-
-typedef struct clh_qnode {
-    clh_qnode_state_t value;
-
-    PAD_TO_NEXT_CACHE_LN(sizeof(clh_qnode_state_t));
-} clh_qnode_t;
-
-typedef struct clh_qnode_p {
-    clh_qnode_t *node;
-    clh_qnode_t *next;
-    /* This is the software IPI flag */
+} clh_req_state_t;
+
+/* Lock request */
+typedef struct clh_req {
+    clh_req_state_t state;
+} ALIGN(L1_CACHE_LINE_SIZE) clh_req_t;
+
+/* Our node (called "Process" in the paper) */
+typedef struct clh_node {
+    clh_req_t *watch; // Used by predecessor to grant the lock to us.
+    clh_req_t *myreq; // Used to grant the lock to our successor.
+    /* This is the software blocking IPI flag */
     word_t ipi;
-
-    PAD_TO_NEXT_CACHE_LN(sizeof(clh_qnode_t *) +
-                         sizeof(clh_qnode_t *) +
-                         sizeof(word_t));
-} clh_qnode_p_t;
+} ALIGN(L1_CACHE_LINE_SIZE) clh_node_t;
```

Comment:
I wonder, does this struct really have an alignment requirement? The member have one that should impose a certain placement of the struct.

Reply:
Not really, it avoids some cache line bouncing. All data in `clh_node_t` is per-node, the alignment is to avoid cache line sharing between nodes.

So it's not required for correctness, only for performance, and the alignment applies to the whole of `clh_node_t` and not individual members.

Reply:
Interesting. I didn't know that the aligned attribute padded the internal size of structs.  @lsf37, how will this play with the CParser?

Reply:
Makes sense.

I used ALIGNED because I really don't like `PAD_TO_NEXT_CACHE_LN`, which is also very error prone. I can undo that part of the change if necessary of course.

Reply:
> Interesting. I didn't know that the aligned attribute padded the internal size of structs. @lsf37, how will this play with the CParser?

I don't think it will be able to deal with that. It might accept it, but if we actually want to verify anything about them, we will need structs to be packed (i.e. no compiler-generated padding). Since this is not a verified config, it's probably fine to use this for now and deal with making padding explicit at some later point.

Reply:
Mind that the only critical alignment and padding in the code is the one for EXCL_RES_GRANULE_SIZE, the others are only for performance. Manual padding structures does not guarantee alignment for over-aligned data either, that can only be verified on the binary, after linking.